Provide JSON Schema for configuration files

XVincentX opened the pull request:
Connect #575 
Closes #575
Closes #415

This pull request will introduce JSON Schemas for configuration files. This should prevent a lot of troubles when trying to start the gateway.

Unfortunately, it is not complete and it is not possible to make it so complete, as we have some hermaphrodite parts that will have to be solved first. It's a good first step, though.

DrMegavolt commented:
I saw the comment that `validateGatewayConfig` cannot be done . then why we need this logic that does the same with reduced functionality ? 

validateGatewayConfig has mixed functionality. some code is validating. some setting defaults
some allows different format
validating and defaults can be moved to the schema

XVincentX commented:
> validateGatewayConfig has mixed functionality. some code is validating. some setting defaults
some allows different format validating and defaults can be moved to the schema

I've moved everything I could from that two functions. There's no more validation or default values in the functions now.

What is left is simply [these single object to array transformations fallbacks](https://github.com/ExpressGateway/express-gateway/blob/feat/config-json-schema/lib/gateway/pipelines.js#L164-L166) and some [runtime checks with regards of other parts of the gateway](https://github.com/ExpressGateway/express-gateway/blob/feat/config-json-schema/lib/gateway/pipelines.js#L208-L210) that a JSON Schema cannot do.

I'd love to get rid of these but that would be a breaking change — so we can't make it now I guess.

>  then why we need this logic that does the same with reduced functionality?

JSON Schema is doing way more than the current logic we have in place. It's checking required parameters, their types, eventually dependencies and so on. We gain a lot more than what we have.

DrMegavolt commented:
do you need https://github.com/ExpressGateway/express-gateway/blob/feat/config-json-schema/lib/gateway/pipelines.js#L158-L160 ? 

and after that validateGatewayConfig is actually a `normalizeGatewayConfig`

DrMegavolt commented:
and for policiesConfig.indexOf(policyNames[0])

isn't it just a enum validation ?  
so name can be one of registered policy.

DrMegavolt commented on lib/gateway/pipelines.js:
```diff
@@ -237,10 +205,6 @@ function validatePipelinePolicies (policies, config) {
-      throw new Error('there should be one and only one policy per policy-object in pipeline configuration');
-    }
-
     if (policiesConfig.indexOf(policyNames[0]) === -1) {
```
and pls replace this part with schema

XVincentX replied:
I am not completely sure whether this is a possible scenario.

Basically you're asking to validate that field based on an enum whose possible fields aren't part of the schema, but part of the payload itself. (Not sure if clear).

I'll try to research to see if it's possible, but at first glance I do not think it is.

DrMegavolt replied:
ok, we can skip this part
$data seems to be not ready yet https://github.com/json-schema-org/json-schema-spec/issues/331

DrMegavolt commented on lib/config/schemas/gateway.config.json:
```diff
@@ -0,0 +1,108 @@
+          "host": {
+            "type": "string"
+          },
+          "paths": {
```
pathRegex is also valid

DrMegavolt replied:
cool, I see pathRegex, 
I also remember `verbs` (or should we call it methods ) so we can skip it for now

XVincentX replied:
It's `methods` and you said it's undocumented. Do you want me to add it?

kevinswiber replied:
Why is `methods` undocumented?

XVincentX replied:
I don't know guys, I wasn't part of the discussion.

DrMegavolt commented on lib/config/schemas/gateway.config.json:
```diff
@@ -0,0 +1,108 @@
+          },
+          "policies": {
+            "type": "array",
+            "items": {}
```
is there a way to describe wildcard policy name and then array of condition\action pairs ?

XVincentX replied:
Yes. I'll modify the schema so policies gets typed. 👍

XVincentX replied:
Actually I recall why I did not define this: as condition/action is in an array:
```
{[condition: {}, action: {}]}
```

this can be a bit tricky to accomplish. I'll see what I can do here.

DrMegavolt commented on lib/config/schemas/system.config.json:
```diff
@@ -0,0 +1,107 @@
+        }
+      },
+      "required": [
+        "cipherKey",
```
if you have defaults, are they required ?

DrMegavolt replied:
And does it mean that we can comment out those values in system.config ?

XVincentX replied:
> if you have defaults, are they required?

Well I'd say yes. They're required fields, but we'll provide good defaults in case you do not provide a value. This is useful when somebody is giving an invalid payload such as 

```javascript
const object = { field: undefined }
``` 

In this case, defaults won't happen (because the field has been provided) but the `required` constraint will mark it as invalid.

> And does it mean that we can comment out those values in system.config ?

Yes. If you have all the default values in place you can technically run the gateway without any config file at all.

DrMegavolt replied:
aha, now I understand about undefined. 

So can we update system config (in `lib` and in CLI templates) to comment out all values that have defaults ?
So it will look like standard apache\nginx\squid or whatever config files. A lot of settings are there with brief explanation,  but they all are disabled

XVincentX replied:
You mean like provide the config file with exact the default values but everything commented out, right?

That might sound a good idea. @kevinswiber ?

1.6? Separate release?

kevinswiber replied:
Let's not block this particular PR waiting for the detailed comments.  Make an issue for it, though, and we can try to get it done for 1.6.

XVincentX commented:
> do you need https://github.com/ExpressGateway/express-gateway/blob/feat/config-json schema/lib/gateway/pipelines.js#L158-L160 ?

Right, I've removed it.

> and after that validateGatewayConfig is actually a normalizeGatewayConfig without true\false magic in code

Regarding the name, you're perfectly right. I've renamed the function in 5b3baec.

Regarding return `true/false`, that value is not indicating whether the validation passed or not (and so starting or not the gateway). 

That value is indicating whether continuing the processing the configuration file or not. The gateway will start anyway.

Does that clarify?

DrMegavolt commented on lib/gateway/pipelines.js:
```diff
@@ -186,7 +154,7 @@ function configurePipeline (pipelinePoliciesConfig, config) {
 }
 
-function validateGatewayConfig (config) {
+function normalizeGatewayConfig (config) {
```
can schema require gatewayConfig.pipelines or  gatewayConfig.pipeline ?
 
I just don't understand the reason to leave true\false login

DrMegavolt commented:
@XVincentX for true/false 
I can leave with it for now.

please check comments in schema and we should be good to go

DrMegavolt commented on lib/config/config.js:
```diff
@@ -89,14 +111,14 @@ class Config {
         log.info(`${watchEvent} event on ${name} file. Reloading system config ` +
           `file from ${this.systemConfigPath}`);
-        this.loadSystemConfig();
+        try { this.loadSystemConfig(); } catch (e) { }
```
can we add some log.debug here ?

XVincentX replied:
Sure!

XVincentX commented:
@DrMegavolt Done!

DrMegavolt commented on lib/config/schemas/gateway.config.json:
```diff
@@ -0,0 +1,137 @@
+        "object",
+        "null"
+      ],
+      "additionalProperties": {
```
what about scopes ?

XVincentX replied:
Added in 7fdb0b3

DrMegavolt commented:
Almost there 
see comment Api Endpoint scopes 
and I would comment out all settings that have defaults in config files

I understand that it cannot be default for http/admin port. because no value means disable. for system config it seems to be all defaults

and especially for db section. the config file can contain redis example, so users will not have to go the doc site

XVincentX commented:
> and I would comment out all settings that have defaults in config files

I think it's a good idea. My only concern is whether it's a good idea to put it as part of 1.6 release. This might be a bit hard to swallow for the guys.

My opinion would be to keep the things a little bit more explicit now, and then in 2.0, when we can finally break some things, we can revisit multiple things, such as default config.

@kevinswiber What's your idea?

kevinswiber commented on lib/config/schemas/gateway.config.json:
```diff
@@ -0,0 +1,152 @@
+        },
+        "tls": {
+          "type": "object",
+          "additioanlProperties": {
```
`additioanlProperties` => `additionalProperties`

XVincentX replied:
😱 
Fixed in 8617d37

kevinswiber commented on lib/config/schemas/system.config.json:
```diff
@@ -0,0 +1,107 @@
+      ]
+    },
+    "session": {
+      "type": "object"
```
Can we get more specific here?  We know the set of properties in here, yeah?

DrMegavolt replied:
as part of implementing session providers it can be that the only property we know is name of the provider

DrMegavolt replied:
I suggest to leave it as is

XVincentX replied:
@kevinswiber List of options is [here](https://github.com/expressjs/session#options)

I didn't enumerate them because our config file is listing only some of these and I wanted to understand more.

I'm fine enumerating all of them eventually.
